### PR TITLE
[#151] Pagination on event records

### PR DESCRIPTION
**Pull request by marceloFA**
hi Wes. Here's a pagination feature to the /events route. Let me know if this was implemented as desired. If so, we can expand the feature to /kinchranks and /sum_of_ranks. 

One important detail: I had no records to populate the rankings and test the pagination feature so if you could, please check it's working properly, please <3

### CODE REVIEW

**Comment by euphwes** [comment]
Hey @marceloFA , so sorry I somehow missed this! I can probably take a look soon, but I think there's a way you can test it too. In `commands.py`, there's a utility command ([generate_fake_comp_results](https://github.com/euphwes/cubers.io/blob/master/app/commands.py#L209-L277)) you can use to fill out the leaderboards for the current comp running. I have like a dozen or so hardcoded names in there, but you could tweak that to have programmatically create 100 or 200 users and submit results for them, which would give you enough data to test your pagination stuff.

Thanks for doing this! I'll look more in-depth in the next couple days, I've just been busy with real life and work lately.

**Comment by marceloFA** [comment]
Thanks for the tips, Wes.  

It's completely fine if you take some time to review this, don't worry at all. By the way, merry christmas to you 🎅

**Comment by euphwes** [comment]
@marceloFA I just tried pulling this down and running, but I'm running into some issues already:

```
line 238, in _build_PersonalBestRecord
    username=record.username,
AttributeError: 'UserEventResults' object has no attribute 'username'
```

You can add some bogus results to your local database with the command I mentioned above, and then actually try running this yourself locally. I can come back to this myself later, but I bet the big issues you'll manage to knock out quickly! Thanks again for picking up this task.

File: app/business/rankings.py [hunk 1]
```diff
@@ -227,13 +228,19 @@ def __init__(self, **kwargs):
...
     """ Builds a PersonalBestRecord from the 5-tuple returned from the ordered PB queries below.
     The tuple looks like (user_id, single/average, comp_id, comp_title, username). """
 
-    user_id, result, comp_id, comp_title, username, comment, user_is_verified = query_tuple
-    return PersonalBestRecord(personal_best=result, user_id=user_id, username=username, comp_id=comp_id,
-        comp_title=comp_title, comment=comment, user_is_verified=user_is_verified)
+    return PersonalBestRecord(
+        personal_best=record.single,
+        user_id=record.user_id,
+        username=record.username,
```
Comment (euphwes) [inline app/business/rankings.py:238]: This is where it's dying when I try to spin up. The record you're passing in doesn't have a `username` attribute.

File: app/business/rankings.py [hunk 2]
```diff
@@ -341,14 +350,14 @@ def get_ordered_pb_averages_for_event(event_id):
         filter(UserEventResults.is_blacklisted.isnot(True)).\
         group_by(UserEventResults.id, UserEventResults.user_id, UserEventResults.average, Competition.id, Competition.title, User.username, User.is_verified).\
         order_by(UserEventResults.id.desc()).\
-        values(UserEventResults.user_id, UserEventResults.average, Competition.id, Competition.title, User.username, UserEventResults.comment, User.is_verified)
```
Comment (euphwes) [inline app/business/rankings.py:344]: So when you remove this line, you're changing the query to just return a list of `UserEventResults` records. The `values(...)` line says to select the specific values you're specifying.

The SQL for this looks something like

```
SELECT <all those fields I specified>
FROM UserEventResults
INNER JOIN User on UserEventResults.user_id = User.id
INNER JOIN CompetitionEvent ...
INNER JOIN Event ...
```

but removing the `value(...)` line basically just only selects the entire `UserEventResults` record and nothing else from the other tables that are joined in the query.

**Review by euphwes — CHANGES_REQUESTED** [review]

**Comment by marceloFA** [comment]
Hi, Wes. Recently I've quite busy with academic tasks so I haven't finished this yet.  My semester is coming to an end in February and then I'll come back to this PR. Thanks for the review and sorry for keeping this open for a while more!

**Comment by euphwes** [comment]
No worries and no hurry! School comes first, for sure.

**Comment by euphwes** [comment]
Oops, somehow merged this when I didn't intend to. I'll fix my commit history and force push, and then see if I can re-open this. @marceloFA

**Comment by euphwes** [comment]
@marceloFA Ok I fixed accidentally pushing this code (and therefore merging this pull request), but I can't seem to re-open it. Sorry about that! Whenever your school semester is over and if you still want to pick this back up, you'll need to open another PR.

**Comment by marceloFA** [comment]
No problem, Wes! :D

**Pull request merged**